Conceptos preliminares:

Optimos locales

Optimos Globales

Funcion objetivo 

Restricciones

Problema general

Conjunto factible

# Optimización Lineal

Un problema de optimizacion lineal en *formato estandar* tiene los siguientes elementos:



- Un vector de $h$ variables de decisión:
   $$x=(x_1,...,x_h)^t$$

- Un vector de $h$ coeficientes de las variables de decision:

$$c=(c_1,...,c_h)^t$$

- Una funcion objetivo lineal:

$$f(x)= c^t \cdot x = c_1\cdot x_1 + ... + c_h \cdot x_h$$

- Un sistema de $i$ restricciones de igualdad:

 $$A_{eq} \cdot x = b_{eq}$$

 Donde:

 $A_{eq}$ es una matriz de coeficientes de tamaño $i$ x $h$

 $b_{eq}$ es un vector de coeficientes de tamaño $i$ 


 - Un distema de $d$ restricciones de desigualdad

$$A_{ineq} \cdot x \leq b_{ineq}$$

Donde:

$A_{ineq}$ es una matriz de coeficientes de tamaño $d$ x $h$

$b_{ineq}$ es un vector de coeficientes de tamaño $d$


- Un vector de $h$ limites inferiores $l=(l_1,...,l_h)^t$ y otro de $h$ limites superiores $u=(u_1,...,u_h)^t$ para las variables de decision $x=(x_1,...,x_h)^t$

Un problema de optimizacion lineal en **formato estandar** (versión de `scipy`) tiene la siguiente forma simbolica:

$$
\underset{x}{Min} \hspace{0.2cm}\lbrace\hspace{0.2cm}  f(x)=c^t \cdot x  \hspace{0.2cm}\rbrace
$$
$$
\hspace{-0.5cm} \text{sujeto a:} \\[0.2cm]
$$

\begin{align*}
 &\hspace{1cm} A_{eq}\cdot x =b_{eq} \\[0.1cm]
&\hspace{1cm} A_{ineq} \cdot x = b_{ineq} \\[0.1cm]
&\hspace{1cm} l \leq x \leq u
\end{align*}


Observacion:

Este formato estandar es el empleado por la libreria de `Python` que usaremos para resolver estos problemas, denominada `scipy`. Pero e¡no es el formato estandar clásico, aquel en el que se bása el famoso método simplex.

Ejemplo:

$$
\underset{x}{Min} \hspace{0.2cm}\lbrace\hspace{0.2cm}  f(x)= x_1 + 2 x_2 - 5 x_3  \hspace{0.2cm}\rbrace
$$
$$
\hspace{-3cm} \text{sujeto a:} \\[0.2cm]
$$

\begin{align*}
 &\hspace{1cm} 2x_1 + 6x_2 + x_3 = 0 \\[0.1cm]
  &\hspace{1cm} 3x_1 - 2x_3 = -2 \\[0.2cm]

&\hspace{1cm}x_1 + 2x_2 + 3x_3 \leq 8 \\[0.1cm]
&\hspace{1cm}5x_2 -7x_3 \leq 10 \\[0.1cm]
&\hspace{1cm}9x_1 + 10x_2 - 3x_3 \leq -1 \\[0.2cm]
&\hspace{1cm} 0 \leq x_1 \\[0.1cm]
&\hspace{1cm} 2 \leq x_2 \leq 4 \\[0.1cm]
&\hspace{1cm} 7 \leq x_3 
\end{align*}

Por tanto, los elementos de este problema de OL son:

$x=(x_1,x_2,x_3)^t$

$c=(1,2,-5)^t$

$A_{eq} = \begin{pmatrix}
    2 & 6 & 1 \\
    3 & 0 & -2 
    \end{pmatrix}$

$b_{eq}=(0 , -2)^t$

$A_{ineq} = \begin{pmatrix}
    1 & 2 & 3 \\
    0 & 5 & -7 \\
    9 & 10 & -3  
    \end{pmatrix}$

$b_{ineq}=(8 , 10, -1)^t$


Conversión de un problema de optimizacion lineal al formato estandar

Dado un problema de OL que no esta en formato estandar, como por ejemplo:

$$
\underset{x}{Max} \hspace{0.2cm}\lbrace\hspace{0.2cm}  f(x)= 29x_1 + 45 x_2  \hspace{0.2cm}\rbrace
$$
$$
\hspace{-3cm} \text{sujeto a:} \\[0.2cm]
$$

\begin{align*}
 &\hspace{1cm} 2x_1 + 8x_2 + x_3 = 60 \\[0.1cm]
  &\hspace{1cm} 4x_1 + 4x_2 + x_4 = 60 \\[0.2cm]

&\hspace{1cm}x_1 - x_2 - 3x_3 \leq 5 \\[0.1cm]
&\hspace{1cm}2x_1 -3x_2 -7x_3 + 3x_4 \geq 10 \\[0.2cm]

&\hspace{1cm} 0 \leq x_1   \\[0.1cm]
&\hspace{1cm} 0 \leq x_2 \leq 6  \\[0.1cm]
&\hspace{1cm} x_3 \leq 0.5 \\[0.1cm]
&\hspace{1cm} -3 \leq x_4
\end{align*}

Lo podemos convertir al formato estandar a traves de las siguientes operaciones:

- Pasar el problema de maximizar a minimizar:

$$
\underset{x}{Max} \hspace{0.2cm}\lbrace\hspace{0.2cm}  f(x)= 29x_1 + 45 x_2  \hspace{0.2cm}\rbrace = \underset{x}{Min} \hspace{0.2cm}\lbrace\hspace{0.2cm} - f(x)= -29x_1 - 45 x_2  \hspace{0.2cm}\rbrace
$$

- Añadir las variables faltantes a la función  objetivo:


$$\underset{x}{Min} \hspace{0.2cm}\lbrace\hspace{0.2cm} - f(x)= -29x_1 - 45 x_2 + 0\cdot x_3 + 0\cdot x_4  \hspace{0.2cm}\rbrace$$

- Convertir las restricciones de $\geq$ en $\leq$

$$2x_1 -3x_2 -7x_3 + 3x_4 \geq 10 \hspace{0.2cm} \Leftrightarrow\hspace{0.2cm} -2x_1 +3x_2 +7x_3 - 3x_4 \leq -10$$

- Obtenemos los elementos del problema estandar:

$$x=(x_1,x_2,x_3,x_4)^t$$

$$c=(-29,-45,0,0)^t$$

$$A_{eq}= \begin{pmatrix}
    2 & 8 & 1 & 0\\
    4 & 4 & 0 & 1 
    \end{pmatrix}$$

$$b_{eq}=\begin{pmatrix}
    60 \\
    60 
    \end{pmatrix}$$

$$A_{ineq} = \begin{pmatrix}
    1 & -1 & -3 & 0\\
    -2 & 3 & 7 & -3 
    \end{pmatrix}$$

$$b_{ineq} = \begin{pmatrix}
    5 \\
    -10 
    \end{pmatrix}$$

Por tanto, el problema expresado en formato estandar es:

$$
\underset{x}{Min} \hspace{0.2cm}\lbrace\hspace{0.2cm}  -f(x)= -29x_1  -45 x_2 + 0 x_3 + 0 x_4  \hspace{0.2cm}\rbrace
$$
$$
\hspace{-3cm} \text{sujeto a:} \\[0.2cm]
$$

\begin{align*}
 &\hspace{1cm} 2x_1 + 8x_2 + x_3 = 60 \\[0.1cm]
  &\hspace{1cm} 4x_1 + 4x_2 + x_4 = 60 \\[0.2cm]

&\hspace{1cm}x_1 - x_2 - 3x_3 \leq 5 \\[0.1cm]
&\hspace{1cm}-2x_1 +3x_2 +7x_3 - 3x_4 \leq -10 \\[0.2cm]

&\hspace{1cm} 0 \leq x_1   \\[0.1cm]
&\hspace{1cm} 0 \leq x_2 \leq 6  \\[0.1cm]
&\hspace{1cm} x_3 \leq 0.5 \\[0.1cm]
&\hspace{1cm} -3 \leq x_4
\end{align*}

In [2]:
import numpy as np
from scipy.optimize import linprog

In [9]:
c = np.array([-29.0, -45.0, 0.0, 0.0])

A_ineq = np.array([[1.0, -1.0, -3.0, 0.0],
                [-2.0, 3.0, 7.0, -3.0]])

b_ineq = np.array([5.0, -10.0])

A_eq = np.array([[2.0, 8.0, 1.0, 0.0],
                [4.0, 4.0, 0.0, 1.0]])

b_eq = np.array([60.0, 60.0])

x1_bounds = (0, None)
x2_bounds = (0, 6)
x3_bounds = (-np.inf, 0.5)  # +/- np.inf can be used instead of None
x4_bounds = (-3.0, None)

bounds = [x1_bounds, x2_bounds, x3_bounds, x4_bounds]
result = linprog(c, A_ub=A_ineq, b_ub=b_ineq, A_eq=A_eq, b_eq=b_eq, bounds=bounds)

print(result)

           con: array([0., 0.])
 crossover_nit: 0
         eqlin:  marginals: array([-2.88717949, -5.41538462])
  residual: array([0., 0.])
           fun: -505.974358974359
       ineqlin:  marginals: array([-5.17435897, -1.80512821])
  residual: array([0., 0.])
         lower:  marginals: array([0., 0., 0., 0.])
  residual: array([9.41025641, 5.17948718,        inf, 4.64102564])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 3
         slack: array([0., 0.])
        status: 0
       success: True
         upper:  marginals: array([0., 0., 0., 0.])
  residual: array([       inf, 0.82051282, 0.75641026,        inf])
             x: array([ 9.41025641,  5.17948718, -0.25641026,  1.64102564])


x: array([ 9.41025641,  5.17948718, -0.25641026,  1.64102564]) is the values of the decision variables that minimizes the objective function while satisfying the constraints.

 fun: -505.974358974359 is the optimal value of the objective function c @ x

In [13]:
c @ result.x 

-505.974358974359

Si el problema no tiene solución esto nos saldrá como resultado. Por ejemplo, si al problema anterior cambiamos la restriccion $ 0 \leq x_2 \leq 6$ por  $0 \leq x_2 \leq 5$ , el problema no tiene solución:

In [14]:
c = np.array([-29.0, -45.0, 0.0, 0.0])

A_ineq = np.array([[1.0, -1.0, -3.0, 0.0],
                [-2.0, 3.0, 7.0, -3.0]])

b_ineq = np.array([5.0, -10.0])

A_eq = np.array([[2.0, 8.0, 1.0, 0.0],
                [4.0, 4.0, 0.0, 1.0]])

b_eq = np.array([60.0, 60.0])

x1_bounds = (0, None)
x2_bounds = (0, 5)
x3_bounds = (-np.inf, 0.5)  # +/- np.inf can be used instead of None
x4_bounds = (-3.0, None)

bounds = [x1_bounds, x2_bounds, x3_bounds, x4_bounds]
result = linprog(c, A_ub=A_ineq, b_ub=b_ineq, A_eq=A_eq, b_eq=b_eq, bounds=bounds)

print(result)

           con: None
 crossover_nit: 0
         eqlin:  marginals: None
  residual: None
           fun: None
       ineqlin:  marginals: None
  residual: None
         lower:  marginals: None
  residual: None
       message: "The problem is infeasible. (HiGHS Status 8: model_status is Infeasible; primal_status is b'At lower/fixed bound')"
           nit: 3
         slack: None
        status: 2
       success: False
         upper:  marginals: None
  residual: None
             x: None
